File: Functions/Controller.m
classdef Controller
    %UNTITLED2 Summary of this class goes here
    %   Detailed explanation goes here
    
    % Channels
    
    % 2         4
    % ^         ^ 
    % |         |
    % L _ > 1   L _ > 3
    
    properties
        
        joy
        
        dir
        
        ch
        deadband
        
        pressed
        lastPressed
        
    end
    
    methods
        
        %Setup
        function this = Controller(joy)
            
            this.joy = vrjoystick(joy);
            
            this.dir = -1;
            
            this.ch = zeros(1,13);
            this.deadband = 0.35;
            
            this.pressed = zeros(1,13);
            this.lastPressed = zeros(1,13);
            
        end
        
        
        %Update
        function this = updateController(this)
            
            this.lastPressed = this.pressed;
            [this.ch, this.pressed, this.dir] = read(this.joy);
            
            this.ch = this.ch .* (abs(this.ch) > this.deadband) .* [1 -1 1 -1];
            
        end
        
        %Setters
        
        function this = setDeadband(this,deadband)
            this.deadband = deadband;
        end
        
        %Getters
        
        function state = dirPad(this)
            state = this.dir;   
        end
        
        function state = channel(this,num)
            state = this.ch(num);
        end
        
        function state = button(this,num)
            state = this.pressed(num);
        end
        
        function state = buttonPressed(this,num)
            state = this.pressed(num) & ~this.lastPressed(num);
        end
        
        %Tools
        
        function value = adjustValue(this,up,down,value,dV)
            if this.buttonPressed(up)
                value = value + dV;
            elseif this.buttonPressed(down)
                value = value - dV;
            end
        end
            
        
    end
    
end

File: Functions/Controller.py
import threading
from inputs import get_gamepad # pip install inputs
import numpy as np


class Controller(threading.Thread):
    #UNTITLED2 Summary of self class goes here
    #   Detailed explanation goes here

    # Channels

    # 2         4
    # ^         ^
    # |         |
    # L _ > 1   L _ > 3


    deadband = None
    ch = [0,0,0,0]

    pressed = None
    lastPressed = None
    
    #Setup
    def __init__(self, *args):

        self.dir = -1

        self.deadband = 0.35

        self.pressed = np.zeros(13)
        self.lastPressed = np.zeros(13)
        self.running = True
        super().__init__(*args)
        self.start()

    def run(self):
        while self.running:
            try:
                #self.ch = [0,0,0,0]
                events = get_gamepad()
                for event in events:
                    if not event.ev_type == 'Sync' and not event.ev_type == 'Misc':
                        if event.code == 'BTN_TRIGGER':
                            self.pressed[0] = event.state
                        elif event.code == 'BTN_THUMB':
                            self.pressed[1] = event.state
                        elif event.code == 'BTN_THUMB2':
                            self.pressed[2] = event.state
                        elif event.code == 'BTN_TOP':
                            self.pressed[3] = event.state
                        elif event.code == 'BTN_TOP2':
                            self.pressed[4] = event.state
                        elif event.code == 'BTN_PINKIE':
                            self.pressed[5] = event.state
                        elif event.code == 'BTN_BASE':
                            self.pressed[6] = event.state
                        elif event.code == 'BTN_BASE2':
                            self.pressed[7] = event.state
                        elif event.code == 'BTN_BASE3':
                            self.pressed[8] = event.state
                        elif event.code == 'BTN_BASE4':
                            self.pressed[9] = event.state
                        ## Left joypad
                        elif event.code == 'ABS_Y':
                            self.ch[1] = 128 - event.state
                        elif event.code == 'ABS_X':
                            self.ch[0] = event.state - 128
                    else:
                            pass
            
            except inputs.UnpluggedError:
                pass
    #Update
    def updateController(self):
        self.lastPressed = self.pressed

    #Setters

    def setDeadband(self, deadband):
        self.deadband = deadband

    #Getters

    def channel(self, num):
        return self.ch[num]

    def button(self, num):
        return self.pressed(num)

    def buttonPressed(self, num):
        return self.pressed(num) & ~self.lastPressed(num)

    #Tools

    def adjustValue(self,up,down,value,dV):
        if self.buttonPressed(up):
            value = value + dV
        elif self.buttonPressed(down):
            value = value - dV
if __name__ == '__main__':
    c = Controller()
